examples/hybrid_jobs/6_QNSPSA_optimizer_with_embedded_simulator/source_scripts/QNSPSA.py

```diff
@@ -0,0 +1,246 @@
+import random
+import pennylane as qml
+from pennylane import numpy as np
+from scipy.linalg import sqrtm
+
+
+class QNSPSA:
+    """Quantum natural SPSA optimizer. Refer to https://quantum-journal.org/papers/q-2021-10-20-567/
+    for a detailed description of the methodology. When disable_metric_tensor
+    is set to be True, the metric tensor estimation is disabled, and QNSPSA is
+    reduced to be a SPSA optimizer.
+
+    Args:
+        lr (float): The learn rate.
+        regularization (float): Regularitzation term to the Fubini-Study
+            metric tensor for numerical stability.
+        finite_diff_step (float): step size to compute the finite difference
+            gradient and the Fubini-Study metric tensor.
+        resamplings (int): The number of samples to average for each parameter
+            update.
+        blocking (boolean): When set to be True, the optimizer only accept
+            updates that leads to a loss value no larger than the loss value
+            before update, plus a tolerance. The tolerance is set with the
+            parameter history_length.
+        history_length (int): When blocking is True, the tolerance is set to be
+            the average of the cost values in the last history_length steps.
+        disable_metric_tensor (boolean): When set to be True, the optimizer is
+            reduced to be a (1st-order) SPSA optimizer.
+        seed (int): Seed for the random sampling.
+    """
+
+    def __init__(
+        self,
+        lr=1e-3,
+        regularization=1e-3,
+        finite_diff_step=1e-2,
+        resamplings=1,
+        blocking=True,
+        history_length=5,
+        disable_metric_tensor=False,
+        seed=None,
+    ):
+        self.lr = lr
+        self.reg = regularization
+        self.finite_diff_step = finite_diff_step
+        self.metric_tensor = None
+        self.k = 1
+        self.resamplings = resamplings
+        self.blocking = blocking
+        self.last_n_steps = np.zeros(history_length)
+        self.history_length = history_length
+        self.disable_metric_tensor = disable_metric_tensor
+        random.seed(seed)
+        return
+
+    def step(self, cost, params):
+        """Update trainable arguments with one step of the optimizer.
+
+        .. warning::
+            When blocking is set to be True, use step_and_cost instead, as loss
+            measurements are required for the updates for the case.
+
+        Args:
+            cost (qml.QNode): the QNode wrapper for the objective function for
+            optimization
+            params (np.array): Parameter before update.
+
+        Returns:
+            np.array: The new variable values after step-wise update.
+        """
+        if self.blocking:
+            raise Exception(
```

Comment:
I think this can be a (one-time) warning instead - users might have existing code that doesn't need the actual cost value. Just call step_and_cost and discard the cost in that case.